applications/sintering/include/pf-applications/sintering/residual_wrapper.h

```diff
@@ -0,0 +1,229 @@
+    using VectorType  = vector_type;
+    using BlockVectorType =
+      LinearAlgebra::distributed::DynamicBlockVector<Number>;
+
+    // This mimic the operator's interface
+    template <int with_time_derivative = 2>
+    void
+    evaluate_nonlinear_residual(
+      BlockVectorType &      dst,
+      const BlockVectorType &src,
+      const std::function<void(const unsigned int, const unsigned int)>
+        pre_operation = {},
+      const std::function<void(const unsigned int, const unsigned int)>
+        post_operation = {}) const
+    {
+      static_assert(with_time_derivative == 1 || with_time_derivative == 2);
+
+      if constexpr (with_time_derivative == 2)
+        evaluate_for_residual(dst, src, pre_operation, post_operation);
+      else if (with_time_derivative == 1)
+        evaluate_for_jacobian(dst, src, pre_operation, post_operation);
+    }
+
+  private:
+    virtual void
+    evaluate_for_residual(
+      BlockVectorType &      dst,
+      const BlockVectorType &src,
+      const std::function<void(const unsigned int, const unsigned int)>
+        pre_operation = {},
+      const std::function<void(const unsigned int, const unsigned int)>
+        post_operation = {}) const = 0;
+
+    virtual void
+    evaluate_for_jacobian(
+      BlockVectorType &      dst,
+      const BlockVectorType &src,
+      const std::function<void(const unsigned int, const unsigned int)>
+        pre_operation = {},
+      const std::function<void(const unsigned int, const unsigned int)>
+        post_operation = {}) const = 0;
+  };
+
+  template <typename Number, typename OperatorType>
+  class ResidualWrapperBase : public ResidualWrapper<Number>
+  {
+  public:
+    using value_type      = typename ResidualWrapper<Number>::value_type;
+    using vector_type     = typename ResidualWrapper<Number>::vector_type;
+    using VectorType      = typename ResidualWrapper<Number>::VectorType;
+    using BlockVectorType = typename ResidualWrapper<Number>::BlockVectorType;
+
+    ResidualWrapperBase(const OperatorType &nonlinear_operator)
+      : nonlinear_operator(nonlinear_operator)
+    {}
+
+  protected:
+    template <unsigned int with_time_derivative = 2>
+    void
+    do_evaluate_nonlinear_residual(
+      BlockVectorType &      dst,
+      const BlockVectorType &src,
+      const std::function<void(const unsigned int, const unsigned int)>
+        pre_operation = {},
+      const std::function<void(const unsigned int, const unsigned int)>
+        post_operation = {}) const
+    {
+      if constexpr (has_evaluate_nonlinear_residual_with_pre_post<
+                      OperatorType,
+                      BlockVectorType>)
+        nonlinear_operator
+          .template evaluate_nonlinear_residual<with_time_derivative>(
+            dst, src, pre_operation, post_operation);
+      else
+        nonlinear_operator
+          .template evaluate_nonlinear_residual<with_time_derivative>(dst, src);
```

Comment:
I think you need to perform pre/post here.

Reply:
Do you mean that we need to execute `pre_operation()` and `post_operation()` manually via separate `MyMatrixFreeTools::cell_loop_wrapper()` calls with some dummy `cell_operation` if the provided `nonlinear_operator` does not support pre/post? Of course in the case `pre_operation()` and/or `post_operation()` are not empty.

Reply:
If `pre_operation` or `post_operation` are non-empty, they need to be executed. For now, we can assert that they are empty?

Reply:
Yea, I see the point. What about 8691efb0a8d3d2ab56dda72f5dc4fa704185dfe5 accompanied by f9d8b7bd3febd45a3d38067eb18ab972edc2199e?